puzzle-utils/words/ladder.py

```diff
@@ -0,0 +1,148 @@
+from .search import search
+
+def ladder(word):
+    """
+    Given a word, find all "adjacent" words in the dictionary
+    i.e. all words with edit distance of 1 from the given word
+
+    Params:
+    word (required) - The base word
+
+    Alternatives:
+    find_ladders        - Find all ladders between two words
+    find_ladders_pretty - Print all unique ladders between two words nicely
+    """
+    found_words = set()
+    for i in range(len(word)):
+        pattern = list(word)
+        pattern[i] = '.'
+        search_results = search("^" + "".join(pattern) + "$")
+        for result in search_results:
+            if result != word and result not in found_words:
+                found_words.add(result)
+    return found_words
+
+def edit_distance(left_word, right_word):
+    """
+    This assumes the lengths of the words are the same.
+    The result is the edit distance of the two words (number of
+    substitutions required to tranform left_word into right_word).
+    """
+    if len(left_word) != len(right_word):
+        raise ValueError("Word ladder words must be same length")
+
+    distance = 0;
+    for i in range(len(left_word)):
+        if left_word[i] != right_word[i]:
+            distance += 1
+    return distance
+
+def follow_ladder_path(word_path, right_word, cached_searches, depth, exact_depth, max_depth, use_heuristic):
+    """
+    Helper to recursively find a word ladder path
+    """
+    # We're too deep, this will take too long
+    if depth > max_depth:
+        return None
+
+    # We don't care about this, we want a specific depth
+    if exact_depth is not None and depth > exact_depth:
+        return None
+
+    # Find words closer to destination (avoiding cycles)
+    current_word = word_path[-1]
+    max_edit_distance = edit_distance(current_word, right_word) if use_heuristic else len(current_word)
+    adjacent_words = []
+    if current_word in list(cached_searches.keys()):
+        adjacent_words = cached_searches[current_word]
+    else:
+        adjacent_words = list(ladder(current_word))
+        cached_searches[current_word] = adjacent_words
+    next_words = list(filter(lambda w: w not in word_path and edit_distance(w, right_word) <= max_edit_distance, adjacent_words))
+
+    # No ladder exists this way
+    if len(next_words) == 0:
+        return None
+
+    # We found a ladder! Append if it's the right depth
+    if right_word in next_words:
+        if exact_depth is None or depth == exact_depth:
+            found_path = list(word_path)
+            found_path.append(right_word)
+            return found_path
+
+    found_paths = []
+    for next_word in next_words:
+        results = None
+        next_path = list(word_path)
+        next_path.append(next_word)
```

Comment:
You don't have to create a new list here. You can append `next_word`, do the recursive call, and remove it afterwards.

Reply:
So the problem I found while writing this was that mutable objects in a recursive function in Python are effectively passed by reference to the next iteration. I'm not sure what the implementation actually looks like, but I came across loads of problems because I was having a hard time freely exploring different paths without shooting myself in the foot since I didn't know this was something that happens in Python.

The approach I took to get this running is to make deep copies of lists when I wanted to be able to manipulate them. I may have even gotten overzealous and done that in some places that didn't need it. I'll have another look at this to see if I can do this better with the way Python lists work in mind